lib/ABI/ApplyRegisterStateDeductions.cpp

```diff
@@ -0,0 +1,292 @@
+
+template<typename CallingConvention,
+         bool EnsureABIConformance = false,
+         typename AccessorType = void>
+void singlePositionBasedDeduction(LocalState First,
+                                  LocalState Second,
+                                  bool &IsRequired,
+                                  RegisterStateMap &State,
+                                  const AccessorType &Accessor) {
+  constexpr model::ABI::Values CurrentABI = CallingConvention::ABI;
+
+  // If `State` is `nullptr` it's not accounted for, hence we see it as `Maybe`.
+  bool FirstYesOrDead = First.State != nullptr ?
+                          model::RegisterState::isYesOrDead(*First.State) :
+                          false;
+  bool SecondYesOrDead = Second.State != nullptr ?
+                           model::RegisterState::isYesOrDead(*Second.State) :
+                           false;
+
+  if (First.State != nullptr && Second.State == nullptr) {
+    // Second is not mentioned, set First as `YesOrDead` if it isn't already.
+    if (IsRequired && !FirstYesOrDead) {
+      *First.State = model::RegisterState::YesOrDead;
+      FirstYesOrDead = true;
+    }
+  } else if (First.State == nullptr && Second.State != nullptr) {
+    // First is not mentioned, set Second as `YesOrDead` if it isn't already.
+    if (IsRequired && !SecondYesOrDead) {
+      *Second.State = model::RegisterState::YesOrDead;
+      SecondYesOrDead = true;
+    }
+  }
+
+  static Logger Log("ABI-RegisterStateDeduction-PositionBased");
+  if (FirstYesOrDead && !SecondYesOrDead) {
+    if (Second.State != nullptr)
+      *Second.State = model::RegisterState::No;
+    else
+      Accessor(State[Second.Register]) = model::RegisterState::No;
+
+    IsRequired = true;
+  } else if (!FirstYesOrDead && SecondYesOrDead) {
+    if (First.State != nullptr)
+      *First.State = model::RegisterState::No;
+    else
+      Accessor(State[First.Register]) = model::RegisterState::No;
+
+    IsRequired = true;
+  } else if (FirstYesOrDead && SecondYesOrDead) {
+    revng_log(Log,
+              "Impossible to differenciate which one of the two registers "
+              "(`model::Register::"
+                << model::Register::getName(First.Register)
+                << "` and `model::Register::"
+                << model::Register::getName(Second.Register)
+                << "`) should be used: both are `YesOrDead`. (the ABI is "
+                << model::ABI::getName(CurrentABI) << ")");
+    IsRequired = true;
+  } else {
+    if (IsRequired)
+      revng_log(Log,
+                "Impossible to differenciate which one of the two registers "
+                "(`model::Register::"
+                  << model::Register::getName(First.Register)
+                  << "` and `model::Register::"
+                  << model::Register::getName(Second.Register)
+                  << "`) should be used: neither is `YesOrDead` (the ABI is "
+                  << model::ABI::getName(CurrentABI) << ")");
```

Comment:
Usually we try to leave out the brackets only if the next statement is one line.